### PR TITLE
support backup of mariadb using mariabackup

### PR DESCRIPTION
**Pull request by pstorz**
### Thank you for contributing to the Bareos Project!

MariaDB has the tool "mariabackup" which offers the same functionality as percona xtrabackup.
This PR introduces a new filedaemon plugin that can backup mariadb servers via mariabackup.
The plugin is based on the percona-xtrabackup plugin.

Also, this PR adds a systemtests that tests the full backup-restore cycle for mariadb and also adds
this functionality to the percona xtrabackup plugin system tests.

#### Please check

- [x] Short description and the purpose of this PR is present _above this paragraph_
- [x] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General

- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Separate commit for this PR in the CHANGELOG.md, PR number referenced is same
- [x] Commit descriptions are understandable and well formatted

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [ ] ~Required documentation changes are present and part of the PR~
       Documentation will be a separate PR
- [x] `bareos-check-sources --since-merge` does not report any problems
- [x] `git status` should not report modifications in the source tree after building and testing

##### Tests

- [x] Decision taken that a system- or unittest is required (if not, then remove this paragraph)
- [x] The decision towards a systemtest is reasonable compared to a unittest
- [x] Testname matches exactly what is being tested
- [x] Output of the test leads quickly to the origin of the fault

### CODE REVIEW

File: core/src/plugins/filed/python/mariabackup/BareosFdPluginMariabackup.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+                100,
+                "plugin_io called with unsupported IOP:" + str(IOP.func) + "\n",
+            )
+            return bRC_OK
+
+    def end_backup_file(self):
+        """
+        Check if dump was successful.
+        """
+        # Usually the xtrabackup process should have terminated here, but on some servers
```
Comment (bruno-at-bareos) [inline core/src/plugins/filed/python/mariabackup/BareosFdPluginMariabackup.py:427]:
```suggestion
        # Usually the mariabackup process should have terminated here, but on some servers
```

File: core/src/plugins/filed/python/mariabackup/bareos-fd-mariabackup.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# Copyright (C) 2015-2020 Bareos GmbH & Co. KG
```
Comment (bruno-at-bareos) [inline core/src/plugins/filed/python/mariabackup/bareos-fd-mariabackup.py:3]:
```suggestion
# Copyright (C) 2015-2021 Bareos GmbH & Co. KG
```

File: debian/bareos-filedaemon-mariabackup-python-plugin.install [hunk 3]
```diff
@@ -0,0 +1,2 @@
+/usr/local/lib64/bareos/plugins/bareos-fd-mariabackup.py*
```
Comment (bruno-at-bareos) [inline debian/bareos-filedaemon-mariabackup-python-plugin.install:1]: /usr/local seems wong here ? Mean the file shouldn't be part of the tree....

File: core/src/plugins/filed/python/mariabackup/bareos-fd-mariabackup.py (hunk 2 above)
Comment (bruno-at-bareos) [inline core/src/plugins/filed/python/mariabackup/bareos-fd-mariabackup.py:3]:
```suggestion
# Copyright (C) 2015-2021 Bareos GmbH & Co. KG
```

File: debian/bareos-filedaemon-mariabackup-python-plugin.install (hunk 3 above)
Comment (bruno-at-bareos) [inline debian/bareos-filedaemon-mariabackup-python-plugin.install:1]: Shouldn't be part of the source tree !

File: docs/pkglists/Fedora_31.x86_64 [hunk 4]
```diff
@@ -46,6 +46,7 @@ x86_64/bareos-filedaemon-ldap-python-plugin
 x86_64/bareos-filedaemon-libcloud-python-plugin
 x86_64/bareos-filedaemon-ovirt-python-plugin
 x86_64/bareos-filedaemon-percona-xtrabackup-python-plugin
+x86_64/bareos-filedaemon-mariabackup-python-plugin
```
Comment (bruno-at-bareos) [inline docs/pkglists/Fedora_31.x86_64:49]: The whole file shouldn't be part of the source tree !

File: docs/pkglists/Fedora_32.x86_64 [hunk 5]
```diff
@@ -46,6 +46,7 @@ x86_64/bareos-filedaemon-ldap-python-plugin
 x86_64/bareos-filedaemon-libcloud-python-plugin
 x86_64/bareos-filedaemon-ovirt-python-plugin
 x86_64/bareos-filedaemon-percona-xtrabackup-python-plugin
+x86_64/bareos-filedaemon-mariabackup-python-plugin
```
Comment (bruno-at-bareos) [inline docs/pkglists/Fedora_32.x86_64:49]: The whole file shouldn't be part of the source tree !

File: docs/pkglists/Fedora_33.x86_64 [hunk 6]
```diff
@@ -46,6 +46,7 @@ x86_64/bareos-filedaemon-ldap-python-plugin
 x86_64/bareos-filedaemon-libcloud-python-plugin
 x86_64/bareos-filedaemon-ovirt-python-plugin
 x86_64/bareos-filedaemon-percona-xtrabackup-python-plugin
+x86_64/bareos-filedaemon-mariabackup-python-plugin
```
Comment (bruno-at-bareos) [inline docs/pkglists/Fedora_33.x86_64:49]: The whole file shouldn't be part of the source tree !

File: docs/pkglists/SLE_15.x86_64 [hunk 7]
```diff
@@ -46,6 +46,7 @@ x86_64/bareos-filedaemon-ldap-python-plugin
 x86_64/bareos-filedaemon-libcloud-python-plugin
 x86_64/bareos-filedaemon-ovirt-python-plugin
 x86_64/bareos-filedaemon-percona-xtrabackup-python-plugin
+x86_64/bareos-filedaemon-mariabackup-python-plugin
```
Comment (bruno-at-bareos) [inline docs/pkglists/SLE_15.x86_64:49]: The whole file shouldn't be part of the source tree !

File: docs/pkglists/SLE_15_SP1.x86_64 [hunk 8]
```diff
@@ -46,6 +46,7 @@ x86_64/bareos-filedaemon-ldap-python-plugin
 x86_64/bareos-filedaemon-libcloud-python-plugin
 x86_64/bareos-filedaemon-ovirt-python-plugin
 x86_64/bareos-filedaemon-percona-xtrabackup-python-plugin
+x86_64/bareos-filedaemon-mariabackup-python-plugin
```
Comment (bruno-at-bareos) [inline docs/pkglists/SLE_15_SP1.x86_64:49]: The whole file shouldn't be part of the source tree !

File: systemtests/tests/py2plug-fd-mariabackup/testrunner [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+done
+
+}
+
+
+
+shutdown_mysql_server
+
+# kill running mysqld instance listening on our port
+sudo fuser "${test_db_port}"/tcp -k || :
```
Comment (bruno-at-bareos) [inline systemtests/tests/py2plug-fd-mariabackup/testrunner:55]: We discuss to remove it : if build and tests still work 100%

**Review by bruno-at-bareos — CHANGES_REQUESTED** [review]
There's little code changes that need to be addressed and also a few files that shouldn't be there.

File: core/src/plugins/filed/python/mariabackup/BareosFdPluginMariabackup.py [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+
+    def __init__(self, plugindef):
+        # BareosFdPluginBaseclass.__init__(self, plugindef)
+        super(BareosFdMariabackup, self).__init__(plugindef)
+        # we first create and backup the stream and after that
+        # the lsn file as restore-object
+        self.files_to_backup = ["lsnfile", "stream"]
+        self.tempdir = tempfile.mkdtemp()
+        # self.logdir = GetValue(bVarWorkingDir)
+        self.logdir = "/var/log/bareos/"
```
[the remark this answers is not in the document]
Comment (pstorz) [reply core/src/plugins/filed/python/mariabackup/BareosFdPluginMariabackup.py:50]: I checked with other plugins and that is the default value that can be overridden by options.  So no, this path is not set by cmake in any python plugin.

File: debian/bareos-filedaemon-mariabackup-python-plugin.install (hunk 3 above)
Comment (pstorz) [reply debian/bareos-filedaemon-mariabackup-python-plugin.install]: removed

Comment (pstorz) [reply debian/bareos-filedaemon-mariabackup-python-plugin.install]: Exactly, removed

File: docs/pkglists/Fedora_31.x86_64 (hunk 4 above)
Comment (pstorz) [reply docs/pkglists/Fedora_31.x86_64:49]: What do you mean? We added these files to make sure no package is lost accidentially (which already happened)

File: docs/pkglists/Fedora_33.x86_64 (hunk 6 above)
Comment (pstorz) [reply docs/pkglists/Fedora_33.x86_64:49]: These files are checked against what really is built

File: systemtests/tests/py2plug-fd-mariabackup/testrunner (hunk 9 above)
Comment (pstorz) [reply systemtests/tests/py2plug-fd-mariabackup/testrunner]: Yes, that was removed in 0d628428747aa363057fa48951b000a833c13005

**Review by pstorz — COMMENTED** [review]
Adressed all comments and should be OK now.

**Review by bruno-at-bareos — APPROVED** [review]
Everything is now reviewed and ok.
Need to find a committer.

**Pull request merged**